### ISSUE TITLE
Plugin Buider for linux/arm64

### ISSUE THREAD

**Issue by reshout** [krakendio/krakend-ce#655]
Could you please provide krakend/builder for the platform linux/arm64 architecture on Docker Hub?

**Comment by taik0** [comment]
Hi @reshout
The builder is the same for amd64 and arm64 (`krakend/builder` for KrakenD-CE and `krakend/builder-ee` for Krakend-EE)
Since building arm64 binaries needs to be done from real arm64 hardware without emulation it's "easier" to build arm64 binaries using a cross-compiler.

Check the github-action file we use to build the KrakenD-CE binaries:
https://github.com/krakendio/krakend-ce/blob/master/.github/workflows/release.yml#L82-L83

**Comment by reshout** [comment]
Thank you for your kind reply.

I understood that you recommend a cross-compilation.

It would be great if a cross-compilation guide will be added to the plug-in build document.

https://www.krakend.io/docs/extending/writing-plugins/#plugin-builder

**Comment by alombarte** [comment]
Hi @reshout ,

Your request is here, hope this helps:
https://www.krakend.io/docs/extending/writing-plugins/#cross-compiling-plugins-arm64

**Comment by reshout** [comment]
I've followed the guide and confirmed that it works! Thanks a lot.

**Comment by reshout** [comment]
We are going to load our custom plugin on Graviton in AWS.

I made the repository which contains `Makefile` to build the plugin for `linux/arm64` platform.

https://github.com/reshout/krakend-server-example

I have two x86_64 Ubuntu machines. One succeeds in loading the plugin, the other fails.

It fails on the M1 MacBook.

When it works, the log looks like this:

```
$ docker run --platform linux/arm64 --rm -p "8080:8080" -v /home/reshout/workspace/krakend-server-example:/etc/krakend/ devopsfaith/krakend:2.1.4
Parsing configuration file: /etc/krakend/krakend.json
2023/04/21 14:52:41 KRAKEND INFO: Starting KrakenD v2.1.4
2023/04/21 14:52:41 KRAKEND DEBUG: [SERVICE: Plugin Loader] Starting loading process
2023/04/21 14:52:41 KRAKEND DEBUG: [SERVICE: Executor Plugin] plugin #0 (/etc/krakend/plugins/krakend-server-example.so): plugin: symbol ClientRegisterer not found in plugin krakend-server-example
2023/04/21 14:52:42 KRAKEND DEBUG: [PLUGIN: krakend-server-example] Logger loaded
2023/04/21 14:52:42 KRAKEND INFO: [SERVICE: Handler Plugin] Total plugins loaded: 1
2023/04/21 14:52:42 KRAKEND DEBUG: [SERVICE: Modifier Plugin] plugin #0 (/etc/krakend/plugins/krakend-server-example.so): plugin: symbol ModifierRegisterer not found in plugin krakend-server-example
2023/04/21 14:52:42 KRAKEND DEBUG: [SERVICE: Plugin Loader] Loading process completed
```

When it doesn't work, the log looks like this:

```
$ docker run --platform linux/arm64 --rm -p "8080:8080" -v /home/reshout/workspace/krakend-server-example:/etc/krakend/ devopsfaith/krakend:2.1.4
Parsing configuration file: /etc/krakend/krakend.json
2023/04/21 14:58:54 KRAKEND INFO: Starting KrakenD v2.1.4
2023/04/21 14:58:54 KRAKEND DEBUG: [SERVICE: Plugin Loader] Starting loading process
2023/04/21 14:58:54 KRAKEND DEBUG: [SERVICE: Executor Plugin] plugin #0 (/etc/krakend/plugins/krakend-server-example.so): plugin: not implemented
2023/04/21 14:58:54 KRAKEND DEBUG: [SERVICE: Handler Plugin] plugin #0 (/etc/krakend/plugins/krakend-server-example.so): plugin: not implemented
2023/04/21 14:58:54 KRAKEND DEBUG: [SERVICE: Modifier Plugin] plugin #0 (/etc/krakend/plugins/krakend-server-example.so): plugin: not implemented
2023/04/21 14:58:54 KRAKEND DEBUG: [SERVICE: Plugin Loader] Loading process completed
```

I'm not sure how to solve this problem. Can you help me?

**Comment by roy-benshlomo** [comment]
> We are going to load our custom plugin on Graviton in AWS.
> 
> I made the repository which contains `Makefile` to build the plugin for `linux/arm64` platform.
> 
> https://github.com/reshout/krakend-server-example
> 
> I have two x86_64 Ubuntu machines. One succeeds in loading the plugin, the other fails.
> 
> It fails on the M1 MacBook.
> 
> When it works, the log looks like this:
> 
> ```
> $ sha1sum plugins/krakend-server-example.so
> dda4c9016513f6e5e4d897be28aa9dccff0447c2  plugins/krakend-server-example.so
> $ lsb_release -a
> No LSB modules are available.
> Distributor ID:	Ubuntu
> Description:	Ubuntu 20.04.6 LTS
> Release:	20.04
> Codename:	focal
> $ docker run --platform linux/arm64 --rm -p "8080:8080" -v /home/reshout/workspace/krakend-server-example:/etc/krakend/ devopsfaith/krakend:2.1.4
> Parsing configuration file: /etc/krakend/krakend.json
> 2023/04/21 14:52:41 KRAKEND INFO: Starting KrakenD v2.1.4
> 2023/04/21 14:52:41 KRAKEND DEBUG: [SERVICE: Plugin Loader] Starting loading process
> 2023/04/21 14:52:41 KRAKEND DEBUG: [SERVICE: Executor Plugin] plugin #0 (/etc/krakend/plugins/krakend-server-example.so): plugin: symbol ClientRegisterer not found in plugin krakend-server-example
> 2023/04/21 14:52:42 KRAKEND DEBUG: [PLUGIN: krakend-server-example] Logger loaded
> 2023/04/21 14:52:42 KRAKEND INFO: [SERVICE: Handler Plugin] Total plugins loaded: 1
> 2023/04/21 14:52:42 KRAKEND DEBUG: [SERVICE: Modifier Plugin] plugin #0 (/etc/krakend/plugins/krakend-server-example.so): plugin: symbol ModifierRegisterer not found in plugin krakend-server-example
> 2023/04/21 14:52:42 KRAKEND DEBUG: [SERVICE: Plugin Loader] Loading process completed
> ```
> 
> When it doesn't work, the log looks like this:
> 
> ```
> $ sha1sum plugins/krakend-server-example.so
> dda4c9016513f6e5e4d897be28aa9dccff0447c2  plugins/krakend-server-example.so
> $  lsb_release -a
> No LSB modules are available.
> Distributor ID: Ubuntu
> Description:    Ubuntu 22.04.2 LTS
> Release:        22.04
> Codename:       jammy
> $ docker run --platform linux/arm64 --rm -p "8080:8080" -v /home/reshout/workspace/krakend-server-example:/etc/krakend/ devopsfaith/krakend:2.1.4
> Parsing configuration file: /etc/krakend/krakend.json
> 2023/04/21 14:58:54 KRAKEND INFO: Starting KrakenD v2.1.4
> 2023/04/21 14:58:54 KRAKEND DEBUG: [SERVICE: Plugin Loader] Starting loading process
> 2023/04/21 14:58:54 KRAKEND DEBUG: [SERVICE: Executor Plugin] plugin #0 (/etc/krakend/plugins/krakend-server-example.so): plugin: not implemented
> 2023/04/21 14:58:54 KRAKEND DEBUG: [SERVICE: Handler Plugin] plugin #0 (/etc/krakend/plugins/krakend-server-example.so): plugin: not implemented
> 2023/04/21 14:58:54 KRAKEND DEBUG: [SERVICE: Modifier Plugin] plugin #0 (/etc/krakend/plugins/krakend-server-example.so): plugin: not implemented
> 2023/04/21 14:58:54 KRAKEND DEBUG: [SERVICE: Plugin Loader] Loading process completed
> ```
> 
> I'm not sure how to solve this problem. Can you help me?

I have the same issue on M1. I'm trying to run Krakend with a container in version 2.3.1 and the plugin is not loaded.

**Comment by taik0** [comment]
Thank you, @reshout; your test allowed me to find the problem faster. The KrakenD binary has been built with the default  `gold` linker, which looks incompatible with plugins and Alpine (which is no problem in Debian, RedHat, and other Linux distros).
This is fixed in the next release, `v2.3.2`. I've made a PR to your example repo so you can test the changes in the new version.

Thank you for reporting!

**Comment by reshout** [comment]
I've checked the problem is fixed. Thank you for the support!
